### ISSUE TITLE
Create a release Workflow that integrates version-numbering and runs tests before pushing

### ISSUE THREAD

**Issue by ajjackson** [pace-neutrons/Euphonic#326]
Discussion with @oerc0122 

The existing release process for Euphonic is at https://github.com/pace-neutrons/pace-developers/blob/master/euphonic/development/02_releases.md

- There are some manual steps that could be removed/consolidated
- The test_release action always fails at first while waiting for Conda. It can be a bit unclear what this is for and perhaps should be split into test_PyPI and test_conda workflows

As we move to a meson-based build system and rework the corresponding actions, it might be wise to have a single action which:

- Updates version numbers in euphonic.version, CITATION.cff
- Renames the "Unreleased" CHANGELOG section to reflect new version number and creates a new "Unreleased" section
- Commit these changes to master
- Builds wheels with new version number (which cannot be read from git tag at this point) and tests from wheels
- IF builds and tests are successful, create new tag, push to GitHub
- Create GitHub release from CHANGELOG (Github is quite forgiving about editing this later)
- Push wheels to PyPI

**Comment by mducle** [comment]
I think the "Push wheels to PyPI" should be a separate manual action because you can only do it once (PyPI will not allow you to modify a release), so there's a high chance something will go wrong even if you make it through all the stages of the action successfully... I'd stop the automatic release action at the GitHub release stage, then manually test the wheels created in the GitHub release and if they were good then upload it to PyPI...

The system I used with `libpymcr` is to have an [action](https://github.com/pace-neutrons/libpymcr/blob/main/.github/workflows/release.yml#L18) run if there is `RELEASE` in the PR title and the PR is merged. This creates a GitHub release, then it runs the normal pipeline (which runs the test and creates the wheels) but the normal pipeline then [detects](https://github.com/pace-neutrons/libpymcr/blob/main/.github/workflows/main.yml#L184) it is a release and uploads the wheels to the release when it succeeds. The main probably with this workflow is that you have to merge the PR to trigger the release so if the release action has a bug / fails somehow you have to create a new PR. So, you could do something like this but manually triggered using a `workflow_dispatch` which takes the version number as an [input](https://github.blog/changelog/2021-11-10-github-actions-input-types-for-manual-workflows/).

**Comment by ajjackson** [comment]
Our initial thought was also to be wary of the push to PyPI step, but...

Once we've run a test suite on the already-built wheels, what other reassurance are we waiting for? What is the likely failure scenario that doesn't kill the Action first?

**Comment by ajjackson** [comment]
Here's an idea for a 2-step process:

- set_version
  - workflow dispatch takes target branch and version number as input
  - updates files
  - also sets tag

- release
  - workflow dispatch takes target branch
  - if last commit on branch is not tagged, abort
  - if the tag is not a valid version number, abort
  - build wheels and test them
  - if tests succesful, push to PyPI

The key tweaks/improvements here are:
- Simplify releasing from a branch. This makes it easy to backport bugfixes to an older version
- Makes it clear what to do if the build/push step fails: make fixes, move the tag and try again.

We could have the release automatically run on new tags, so that the first attempt at a release is still "one click".

**Comment by ajjackson** [comment]
If the process is more streamlined, we would be more likely to do "release candidate" versions which are another tool for catching packaging problems.

**Comment by ajjackson** [comment]
Another interesting thought for release tests (and maybe regular tests): `uv` is faster than Pip and its pip-like interface includes a `--resolution lowest` option that will automatically find the lowest allowed versions of dependencies. If we used that, we wouldn't need to keep a _minimum_requirements.txt_ file in sync with pyproject.toml.

**Comment by ajjackson** [comment]
One complication: Github deliberately makes it slightly tricky to have actions run in response to an automated push.

https://github.com/peter-evans/create-pull-request/issues/48

It can be done by adding a PAT to secrets, but let's start out with manual triggers and see how it goes.

**Comment by ajjackson** [comment]
A bit of reflection on pre/post releases and CHANGELOG manipulation with @oerc0122 led to this proposal:

- post-releases are releases that make some helpful tweak to packaging or distribution but do not change the main code. By default pip will not "upgrade" a release to a corresponding post-release, but will prefer to install a post-release over the base release.
  - these should have their own CHANGELOG entry, explaining why they were needed.
  - Release logic should expect to find a consistent release-tag, version numbers and changelog references for the post-release

- pre-releases are unsupported builds for development/testing before an official release
  - they are not routinely used in Euphonic development but should be considered when making changes to build systems etc.
  - they should *not* have a CHANGELOG entry: changes remain "Unreleased" until they are offically released.
  - Release logic should expect to find notes in the "Unreleased" section of CHANGELOG and leave this intact if the release tag is a pre-release. Other version numbers should be updated.

- If a release attempt has been cancelled, fixed up and re-attempted, there may already be an appropriate CHANGELOG section.
  - In this case, release logic should make sure nothing is in the "Unreleased" section and continue without making changes. If there is something in "Unreleased", the release pipeline should fail and require manual cleanup.

This keeps the logic for checking and manipulating CHANGELOG fairly simple, and means that as a developer the rule is:

- If fixing a release branch after an unsuccessful release attempt, release notes should go in the (already-generated) section that shares a name with the upcoming release tag
- Otherwise, release notes should always be written in an "Unreleased"

**Comment by ajjackson** [comment]
The updated proposal for a "1-step" release action is:

1. Manual dispatch action, providing the source branch and new version number
2. Validate/normalize the proposed version number
3. update euphonic/version.py and CITATION.cff
4. update CHANGELOG according to principles above
5. create local commit/tag but do not push yet
6. build wheel packages and test with cibuildwheel
7. (in parallel with 6) build sdist and test
8. Push commit and tag to Github
9. Create release payload from CHANGELOG and make Github release
10. Push packages to PyPI

It would make sense to provide a boolean "dry-run" option that quits the workflow after step 7.

Overall then this combines to functions of the existing "build wheels and push to pypi" workflow and the (soon to be merged) "set version" workflow.

I had initially thought that the "set version" workflow would remain separate and be called as a step by "the big one"; but that doesn't make it so easy to defer pushing the commit/tag until tests have run. So maybe a single mega-workflow is the way to go.
